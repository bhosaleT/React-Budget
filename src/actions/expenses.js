import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({ type: "REMOVE_EXPENSE", id });

// EDIT_EXPEPNSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});


/* HOW FETCHING DATA FROM THE DATABASE WORKS FOR THE FIRST TIME 
-- we use functions which will contact firebase and then also push our actions.
-- we will return (dispatch) to get access to dispatch
-- then we reference to database using .ref() and then .once() which then gives us a snapshot which is our data.
-- then we loop through it using forEach and get inside that we get childsnapshot we add the childsnapshot to an array 
   which we will call expenses and then just the id comes form child snapshot and then we spread the childsnapshot.val().
-- Then we will dispatch(setExpenses).
*/


// export const startSetExpenses;
export const startSetExpenses = () => {
  return dispatch => {
   return database.ref('expenses').once('value').then((snapshot)=>{
     const expenses = [];
     snapshot.forEach((childsnapshot) => {
        expenses.push({
          id: childsnapshot.key,
          ...childsnapshot.val()
        });
     });
     dispatch(setExpenses(expenses));
    })
  };
};
