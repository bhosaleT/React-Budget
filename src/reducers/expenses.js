const expensesReducerDefaultState = [];

//This reducer deals with the expenses section of the state the switch cases in this reducer will contain actions to manipulate the expenses state.
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [...state, action.expense];
        case "REMOVE_EXPENSE":
            return state.filter(expense => {
                return action.id !== expense.id;
            });
        case "EDIT_EXPENSE":
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;