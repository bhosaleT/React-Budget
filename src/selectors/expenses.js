//============================= GET VISIBLE EXPENSES ==========================//

/* 
--- WORKING OF FILTERING ---
1.  the function named getVisibleExpenses takes in two things the state.expenses and state.filters
2.  destruct the filter state into individiual variables.
3.  three variables are created to check the startDate, EndDate, text
4.  to check the startDate whether its number or not and then if expense.createdAt is less than or equal to the startDate
5.  to check if the endDate is greater than createdAt
6.  to match the text we use .includes(with the text filter provided) && convert all text to lowercase to make the filtering free or uppercase or lowercase
7.  FILTER all the values where all three are true.
8.  SortBy {using .sort() func} if the sort is based on time then check which ones createdAt timestamp value is lower that one should come first, if its sorted by amount        the larger amount one should come first.
*/

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses
        .filter(expense => {
            const startDateMatch =
                typeof startDate !== "number" || expense.createdAt >= startDate;
            const endDateMatch =
                typeof endDate !== "number" || expense.createdAt <= endDate;
            const desc = expense.description.toLowerCase();
            const textMatch = desc.includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === "date") {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else if (sortBy === "amount") {
                return a.amount > b.amount ? -1 : 1;
            }
        });
};

export default getVisibleExpenses;