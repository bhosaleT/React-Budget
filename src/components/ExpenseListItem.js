import React from "react";
import { removeExpense } from "../actions/expenses";
import { connect } from "react-redux";

/*  ================= WORKING OF DELETE BUTTON =============
-- the mapStateToProps is not always required so to remove a state, we will import our action function removeExpense and call dispatch(removeExpense(with the expense id))
-- $ This code exists in this component because the id is already passed through props and destructured here so we already have it here to be directly passed into the          function.
-- $ Remember the action functions take in objects {}.
*/

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => {
  return (
    <div>
      <h3>{description}</h3>
      <p>Amount: {amount}</p>
      <p>Time: {createdAt}</p>
      <button
        onClick={() => {
          dispatch(removeExpense({ id }));
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default connect()(ExpenseListItem);
