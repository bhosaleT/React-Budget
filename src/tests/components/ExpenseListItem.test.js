import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('Should render a list item', ()=>{
    const wrapper = shallow(<ExpenseListItem expense = {expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
})