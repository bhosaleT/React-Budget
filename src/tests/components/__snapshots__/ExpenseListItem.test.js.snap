// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a list item 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/undefined"
>
  <div>
    <h3
      className="list-item__title"
    />
    <span
      className="list-item__subtitle"
    >
       
      January 1st, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    â‚¹0.00
  </h3>
</Link>
`;
